packages/core/lib/segments/plugins/ec2_plugin.js

```diff
@@ -9,22 +9,86 @@ var EC2Plugin = {
    */
 
   getData: function(callback) {
-    var METADATA_OPTIONS = {
-      host: '169.254.169.254',
-      path: '/latest/dynamic/instance-identity/document'
-    };
-
-    Plugin.getPluginMetadata(METADATA_OPTIONS, function(err, data) {
-      if (err) {
-        logger.getLogger().error('Error loading EC2 plugin: ', err.stack);
-        callback();
-      } else {
-        var metadata = { ec2: { instance_id: data.instanceId, availability_zone: data.availabilityZone }};
-        callback(metadata);
-      }
-    });
+    const TOKEN_PATH = '/latest/api/token';
+    const METADATA_PATH = '/latest/dynamic/instance-identity/document';
+
+    function getMetadataToken() {
+      const TTL = 60; //seconds
+      const tokenOptions = getOptions(TOKEN_PATH, 'PUT', {
+        'X-aws-ec2-metadata-token-ttl-seconds': TTL
+      });
+
+      Plugin.getToken(tokenOptions, function(err, data) {
+        if (err) {
+          logger.getLogger().debug('EC2Plugin failed to get token from IMDSv2. Falling back to IMDSv1.', err.toString());
+          getV1Metadata();
+        } else {
+          // Successfully got token, use it for IMDSv2 metadata request
+          getV2Metadata(data);
+        }
+      });
+    }
+
+    function getV2Metadata(token) {
+      const options = getOptions(METADATA_PATH, 'GET', {
+        'X-aws-ec2-metadata-token': token
+      });
+
+      Plugin.getPluginMetadata(options, function(err, data) {
+        if (err) {
+          logger.getLogger().debug('EC2Plugin failed to get metadata from IMDSv2. Falling back to IMDSv1.', err.toString())
+          getV1Metadata();
+        } else if (data) {
+          const metadata = { ec2: { instance_id: data.instanceId, availability_zone: data.availabilityZone }};
+          callback(metadata);
+        } else {
+          logger.getLogger().debug('Failed to get required EC2 metadata from IMDSv2. Falling back to IMDSv1.');
+          getV1Metadata();
+        }
+      });
+    }
+
+    function getV1Metadata() {
```

Comment:
As far as I can tell, the only difference between v1 and v2 is whether the token is included or not. Can we share the business logic, mostly the URL and the response parsing?

Reply:
Also makes me wonder, is it actually possible for the request to fail? I guess for IDMS v1 it'll just ignore the header? It doesn't hurt to have the fallback logic but it'd be nice to understand what the expected behavior in prod is.

Reply:
Just in case it helps, here's some code in Java I wrote that has less branching. Let me know if it's broken though ;)

https://github.com/aws-samples/aws-xray-sdk-with-opentelemetry-sample/blob/both-otel-and-xray/awsagentprovider/src/main/java/com/softwareaws/xray/opentelemetry/exporters/resources/Ec2ResourcePopulator.java#L64

Reply:
That's a good observation. Really only the token request would fail if we were on a V1-only instance. So I'll refactor to share much more of the logic for the two versions based on your sample. That'll probably make its way to the Java SDK as well :)